File: acwing/蓝桥杯每日一题2025/week3/5539.牛奶交换/5539.牛奶交换.cpp
#include <iostream>
#include <algorithm>
#include <cstring>
#include <cstdio>

using namespace std;

typedef long long LL;

const int N = 4 * 1e5 + 10;
char s[N];
int a[N];

int main()
{
    int n, m;
    scanf("%d%d", &n, &m);
    scanf("%s", s + 1);
    LL sum = 0;
    for (int i = 1; i <= n; i++)
    {
        scanf("%d", a + i);
        sum += a[i];
        a[i + n] = a[i];
    }
    int k = 1;
    while (k < n && s[k] == s[k + 1])
    {
        k++;
    }
    if (k < n)
    {
        for (int i = k + 1; i <= n + k; i++)
        {
            int j = i;
            LL res = 0;
            while (j <= n + k && s[j] == s[i])
            {
                res += a[j];
                j++;
                
            }
            if (s[i] == 'R')
                res -= a[j];
            else
                res -= a[i];

            sum -= min(res, (LL)m);
            i = j - 1;
        }
    }
    printf("%lld", sum);
    return 0;
}

File: acwing/蓝桥杯每日一题2025/week3/5539.牛奶交换/5539.牛奶交换.py
N = 2 * int(1e5) + 10
a = [0] * 2 * N

n, m = map(int, input().split())
s = input()

s = s + s

a = list(map(int, input().split()))

res = sum(a)

a.extend(a)

k = 0
while k < n and s[k] == s[k + 1]:
    k += 1

if k < n:
    i = k + 1
    while i <= n + k:
        j = i
        cnt = 0
        while j <= n + k and s[j] == s[i]:
            cnt += a[j]
            j += 1
        if s[i] == 'R':
            cnt -= a[j - 1]
        else:
            cnt -= a[i]
        res -= min(cnt, m)
        i = j - 1
        i += 1

print(res)
